### ISSUE TITLE
Add ability to specify that direction and other elements should appear on top or bottom staff

### ISSUE THREAD

**Issue by mdgood** [w3c/musicxml#37]
One problematic area for using MusicXML in reflowing or editing applications is that direction, harmony, measure-numbering, and ending elements are tied to a specific staff. Often what is really wanted is to display these elements on the top or bottom staff of a system. Both reflowing and editing can change the music so that a previously hidden staff gets displayed, putting the tempo marking or chord symbol on the incorrect staff.

MusicXML should add the ability to specify that these elements should display on the top or bottom staff of a system. Formatting associated with the element may no longer apply if the reflowed top or bottom staff is no longer the same staff indicated in the original MusicXML file.

See the original MusicXML forum discussion on this and related reflow topics at http://forums.makemusic.com/viewtopic.php?f=12&t=2376&p=6396#p6404.

**Comment by mdgood** [comment]
The DAISY Music Braille Project has pointed out that this is an issue for Braille encoding as well. A MusicXML to Braille translator needs to reliably know when texts and repeats are intended to apply to all parts even if only displayed in a single part in the score.

In some cases, marking appear not only on the top or bottom staff, but on some intermediate staves as well, particularly in scores with a large number of parts. The solution to this issue should be able to indicate that these intermediate staff markings are also part of such an "all staff" marking, not something that is specific to the intermediate staff only.

**Comment by kepper** [comment]
Certain indications (tempo markings etc.) are repeated for every section (brass, woodwinds, strings, …), so there should be a way to describe that situation in MusicXML 3.2.

**Comment by constraint** [comment]
好
----- 原始邮件 -----
发件人：kepper <notifications@github.com>
收件人：w3c/musicxml <musicxml@noreply.github.com>
抄送人：Subscribed <subscribed@noreply.github.com>
主题：Re: [w3c/musicxml] Add ability to specify that direction and other elements should appear on top or bottom staff (#37)
日期：2019年04月04日 21点09分

Certain indications (tempo markings etc.) are repeated for every section (brass, woodwinds, strings, …), so there should be a way to describe that situation in MusicXML 3.2.



&mdash;

**Comment by hhpmusic** [comment]
DAISY Music Braille Project Proposition (examples from German Library For The Blind):

```
<dynamics scope="system"><fff/></dynamics>
<breath-mark scope="staff"/>
<directive scope="system">cresc.</directive>
```

Also, I think there may be an attribute for staff other than the topmost one, letting certain staff or staves display such global texts or symbols, e.g., at the top plus first violin. When the first violin is the topmost staff temporarily, there should be only one copy of such thing to display--this is interpreted by notation softwares anyway.

**Comment by mdgood** [comment]
One way that we could do this is to add a system attribute to the `<direction>`, `<harmony>`, `<ending>`, and `<measure-numbering>` elements. 

For all these elements, the attribute values could be:

- top: This element is intended to be displayed on the top staff of the system.
- also-top: This element is intended to be displayed not only on this part, but also on the top staff of the system. This is intended to handle the situations described by @hhpmusic and @kepper.
- none: There is no system-specific display for this element. This is the default value if the system attribute is not present.

For the `<measure-numbering>` element there could be two additional values:

- bottom: The number is intended to be displayed on the bottom staff of the system.
- also-bottom: The number is intended to be displayed not only on this part, but also on the bottom staff of the system.

I don't think we need these values for the other three elements, but please let me know if you think that is too limiting.

While the `<directive>` element could use this too, that element has been deprecated since MusicXML 2.0. I am inclined not to add any new features to a deprecated element but please let me know if you disagree.

Are there any other elements that could use this attribute? These are the four that are mentioned in the initial issue description, but I don't want to overlook anything.

Adding this attribute to the `<measure-numbering>` element has some implications for #23 that I plan to address separately in that issue.

How does this design look? Are the distinctions clear? Are we handling all the needed cases? Is there a different design that could do this better? Thanks in advance for your comments and suggestions!

**Comment by hhpmusic** [comment]
That's good, thank you Michael!

**Comment by jsawruk** [comment]
I think the names `also-top` and `also-bottom` are confusing. Maybe `part-top`, `system-top`, and `part-and-system-top`? I know those names are verbose, but I think they convey the intent more easily.

**Comment by mdgood** [comment]
@jsawruk Thanks for the suggestion. It would be great to get these value names as clear as possible. I'm not so sure about those values being less confusing though. My main concern is around "system" being repeated both in the attribute name and the attribute value. Can we come up with another clearer alternative? 

Maybe `top-only`, `top-also`, and `none` for the `system` attribute values? We would add `bottom-only` and `bottom-also` for use with the `measure-numbering` element. To see some examples in more context:

`<direction directive="yes" placement="above" system="top-only">` 
`<direction directive="yes" placement="above" system="top-also">`
`<measure-numbering system="bottom-only">measure</measure-numbering>`
`<measure-numbering system="top-also">system</measure-numbering>`

Might that work better, or lead to another idea? I am wondering if perhaps the system attribute should have a different name in the `<measure-numbering>` element, given the overlap with the system value in the element text. Thanks!

**Comment by hhpmusic** [comment]
I suddenly come up with a different approach: Besides the above values, can we also implement the value for certain parts if needed? This comes from the system texts in orchestral scores, which may appear on both top and first violin, sometimes also piano solo part (in concerto). Sibelius can display these texts on certain staves, but the property is still system. So when exporting to Musicxml, we needn't to export duplicated texts, but just export the topmost ones and include the part id's on which they will display.

**Comment by mdgood** [comment]
@hhpmusic, that is what the "top-also" value is meant to convey. It would be used in the first violin and piano parts in your example. The flute (or whatever else is the top instrument) would have the "top-only" value. 

I think this approach is more compatible with earlier versions of MusicXML that what you are suggesting, and both approaches should be able to solve this problem. 

If you think I am missing something, please let me know. Maybe this is related to @jsawruk's point about the names being unclear.

**Comment by jsawruk** [comment]
It seems like there are three states (?):

- Display on this part only: current behavior / `none` value
- Display on the top of the system only: `top`
- Display on this part AND on the top of the system `top-also` / `also-top`

Name suggestions:

- Rename `top` to `only-top`, indicating that this will only display on the top of the system, even though it is specified at the part level.
- Rename `also-top` to `with-top` or `both`, indicating that this will display in both positions.

Another idea might be to just have an attribute specifying only if the element appears on the top system. Part-level visibility could be controlled independently by showing/hiding the element. Thoughts?

**Comment by mdgood** [comment]
Thanks @jsawruk - I think your state analysis is correct, and I like the `only-top` and `with-top` names. I think `with-top` is more clear than `both`, and also lets us specify top or bottom independently for the `<measure-numbering>` element.

I'm not quite sure I am understanding your other idea. The need for `with-top` is for cases like @hhpmusic mentions, where you have a system direction that needs to appear on multiple parts in the score, but only once in a given part. This is something that would be good to be able to import just from the score, rather than checking visibility in each part. If I am missing the point, which I suspect I may be, please let me know!

**Comment by jsawruk** [comment]
Re my other idea: I was trying to see if there was a way to convert this to a binary state for more flexibility. My idea would have covered four possibilities:

- Not visible on part, not visible on system top
- Visible on part, not visible on system top

etc.

However, I wasn't too confident on this idea, and your point about checking visibility on each part outweighs its potential benefit.

I think `only-top` and `with-top` are good names, so let's go with that.

**Comment by hhpmusic** [comment]
When using also or with top, will these texts be placed in other parts
rather than the top one? Then we still get duplicated texts. If
placing it on the topmost part and specify certain part ID's, when
imported, they will appear on correct staves.

Haipeng


On 12/3/20, jsawruk <notifications@github.com> wrote:
> It seems like there are three states (?):
>
> - Display on this part only: current behavior / `none` value
> - Display on the top of the system only: `top`
> - Display on this part AND on the top of the system `top-also` / `also-top`
>
> Name suggestions:
>
> - Rename `top` to `only-top`, indicating that this will only display on the
> top of the system, even though it is specified at the part level.
> - Rename `also-top` to `with-top` or `both`, indicating that this will
> display in both positions.
>
> Another idea might be to just have an attribute specifying only if the
> element appears on the top system. Part-level visibility could be controlled
> independently by showing/hiding the element. Thoughts?
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/w3c/musicxml/issues/37#issuecomment-737630233


-- 
Hu Haipeng
Blind musician and braille music transcriber of China
The BrailleOrch and Open Braille Music Projects
http://www.brailleorch.org

**Comment by mdgood** [comment]
@hhpmusic Instead of part IDs, perhaps we could make use of the optional unique IDs that are available for the `<direction>` element. We could have the content of the `with-top` attribute be a reference to the ID of the corresponding `<direction>` that has an `only-top` attribute.

We will still need developers to use this data to remove potential duplicate texts when importing a MusicXML file, but adding the explicit reference could make that a lot easier.

**Comment by hhpmusic** [comment]
I agree. Thanks!



On 12/4/20, Michael Good <notifications@github.com> wrote:
> @hhpmusic Instead of part IDs, perhaps we could make use of the optional
> unique IDs that are available for the `<direction>` element. We could have
> the content of the `with-top` attribute be a reference to the ID of the
> corresponding `<direction>` that has an `only-top` attribute.
>
> We will still need developers to use this data to remove potential duplicate
> texts when importing a MusicXML file, but adding the explicit reference
> could make that a lot easier.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/w3c/musicxml/issues/37#issuecomment-738514122


-- 
Hu Haipeng
Blind musician and braille music transcriber of China
The BrailleOrch and Open Braille Music Projects
http://www.brailleorch.org

**Comment by mdgood** [comment]
Oops, my previous suggestion would not work. The current proposal is for a `system` attribute for the `<direction>` element with the attribute values `only-top` and `with-top`. In order to add a reference to a unique ID for use with the `with-top` value, we would need to either add another attribute or a child element.

I think I prefer the added attribute. We could add an optional IDREF attribute, say `top-id`, whose value is the `id` of a `<direction>` with an `system="only-top"` attribute value. I think it will be fine to limit this to the `<direction>` element and not include it in the `<measure-numbering>` element.

Is there another way to rearrange things that might work better?

Let me write up a pull request and see how this looks so we can start trying this out.

**Comment by mdgood** [comment]
I've submitted pull request #354 to fix this issue as well as #23. This pull request does not include the idea of a `top-id` attribute. I'd like to experiment and see if we can do without it and still get the results we want.
